engine/classes/Elgg/Database/River.php

```diff
@@ -0,0 +1,318 @@
+		}
+
+		return $items;
+	}
+
+	/**
+	 * Execute the query resolving calculation, count and/or batch options
+	 *
+	 * @return array|\ElggData[]|ElggEntity[]|false|int
+	 * @throws \LogicException
+	 */
+	public function execute() {
+
+		if ($this->options->annotation_calculation) {
+			$clauses = $this->options->annotation_name_value_pairs;
+			if (count($clauses) > 1 && $this->options->annotation_name_value_pairs_operator !== 'OR') {
+				throw new \LogicException("Annotation calculation can not be performed on multiple annotation name value pairs merged with AND");
+			}
+
+			$clause = array_shift($clauses);
+
+			return $this->calculate($this->options->annotation_calculation, $clause->names, 'annotation');
+		} else if ($this->options->count) {
+			return $this->count();
+		} else if ($this->options->batch) {
+			return $this->batch($this->options->limit, $this->options->offset, $this->options->callback);
+		} else {
+			return $this->get($this->options->limit, $this->options->offset, $this->options->callback);
+		}
+	}
+
+	/**
+	 * Build a database query
+	 *
+	 * @param QueryBuilder $qb
+	 *
+	 * @return QueryBuilder
+	 */
+	protected function buildQuery(QueryBuilder $qb) {
+
+		$ands = [];
+
+		foreach ($this->options->joins as $join) {
+			$join->prepare($qb, 'rv');
+		}
+
+		foreach ($this->options->wheres as $where) {
+			$ands[] = $where->prepare($qb, 'rv');
+		}
+
+		$ands[] = $this->buildRiverClause($qb);
+		$ands[] = $this->buildEntityClauses($qb);
+		$ands[] = $this->buildPairedAnnotationClause($qb, $this->options->annotation_name_value_pairs, $this->options->annotation_name_value_pairs_operator);
+		$ands[] = $this->buildPairedRelationshipClause($qb, $this->options->relationship_pairs);
+
+		$ands = $qb->merge($ands);
+
+		if (!empty($ands)) {
+			$qb->andWhere($ands);
+		}
+
+		return $qb;
+	}
+
+	/**
+	 * Process river properties
+	 *
+	 * @param QueryBuilder $qb Query builder
+	 *
+	 * @return CompositeExpression|mixed|null|string
+	 */
+	protected function buildRiverClause(QueryBuilder $qb) {
+		$where = new RiverWhereClause();
+		$where->ids = $this->options->ids;
+		$where->views = $this->options->views;
+		$where->action_types = $this->options->action_types;
+		$where->subject_guids = $this->options->subject_guids;
+		$where->object_guids = $this->options->object_guids;
+		$where->target_guids = $this->options->target_guids;
+		$where->type_subtype_pairs = $this->options->type_subtype_pairs;
+		$where->created_after = $this->options->created_after;
+		$where->created_before = $this->options->created_before;
+
+		return $where->prepare($qb, 'rv');
+	}
+
+	/**
+	 * Add subject, object and target clauses
+	 * Make sure all three are accessible by the user
+	 *
+	 * @param QueryBuilder $qb Query builder
+	 *
+	 * @return CompositeExpression|mixed|null|string
+	 */
+	public function buildEntityClauses($qb) {
+		$ands = [];
+
+		$qb->joinEntitiesTable('rv', 'subject_guid', 'inner', 'se');
```

Comment:
Check if river API allows non-users to be subjects. If not, then we can remove this